### PR TITLE
added gcc09 and vcg04

### PR DESCRIPTION
**Pull request by siddharthlal25**
Added gcc09 and vcg04 as mentioned in #10.

### CODE REVIEW

File: src/color_laws.jl [hunk 1]
```diff
@@ -137,3 +137,39 @@ function cal00_invum(x::Real, Rv::Real)
...
+    x = aa_to_invum(λ)
+    return gcc09_invum(x, Rv)
+end
+
+gcc09(λ::Quantity, Rv::Real = 3.1) = gcc09(ustrip(u"angstrom", λ), Rv) * u"mag"
+
+function gcc09_invum(x::Real, Rv::Real)
+
+    a = zeros(typeof(Rv))
+    b = zeros(typeof(Rv))
```
Comment (giordano) [inline src/color_laws.jl:161]: `zeros` generates an array, I think you wanted to use `zero` instead.  Actually, I don't think these are necessary at all?

File: src/color_laws.jl [hunk 2]
```diff
@@ -137,3 +137,39 @@ function cal00_invum(x::Real, Rv::Real)
...
+    if 3.3 <= x && x <= 11.0  # NUV
```
Comment (giordano) [inline src/color_laws.jl:162]:
```suggestion
    if 3.3 <= x <= 11.0  # NUV
```
:wink:

File: src/color_laws.jl [hunk 3]
```diff
@@ -137,3 +137,39 @@ function cal00_invum(x::Real, Rv::Real)
...
+        a = 1.894 - 0.373 * x - 0.0101 / ((x - 4.57)^2 + 0.0384)
+        b = -3.490 + 2.057 * x + 0.706 / ((x - 4.59)^2 + 0.169)
+    else
+        return 0.0x
+    end
+    if 5.9 <= x && x <= 11.0  # far-NUV
```
Comment (giordano) [inline src/color_laws.jl:168]:
```suggestion
    if 5.9 <= x <= 11.0  # far-NUV
```

File: src/color_laws.jl (hunk 1 above)
Comment (siddharthlal25) [reply src/color_laws.jl:161]: Correct! Thank you!

File: src/color_laws.jl (hunk 2 above)
Comment (siddharthlal25) [reply src/color_laws.jl:162]: Thank you!

File: src/color_laws.jl (hunk 3 above)
Comment (siddharthlal25) [reply src/color_laws.jl:168]: Will keep this in mind from next time! :smiley:

**Comment by mileslucas** [comment]
@giordano any clue why these travis builds are randomly failing?

File: src/DustExtinction.jl [hunk 4]
```diff
@@ -16,6 +18,8 @@ include("dust_maps.jl")
 
 @deprecate ccm89(x::AbstractArray, r_v::Real = 3.1) ccm89.(x, r_v)
 @deprecate od94(x::AbstractArray, r_v::Real = 3.1) od94.(x, r_v)
+@deprecate vcg04(x::AbstractArray, r_v::Real = 3.1) vcg04.(x, r_v)
+@deprecate gcc09(x::AbstractArray, r_v::Real = 3.1) gcc09.(x, r_v)
```
Comment (mileslucas) [inline src/DustExtinction.jl:22]: I wouldn't say you need to add these deprecation statements (notice `cal00` isn't there).

File: src/color_laws.jl [hunk 5]
```diff
@@ -127,13 +127,87 @@ function cal00_invum(x::Real, Rv::Real)
     if x > 1 / 0.12
         return 0.0x
     elseif x > 1 / 0.63
-        k = 2.659 * (((0.011x - 0.198)x + 1.509)x - 2.156)
+        k = 2.659 * Poly([-2.156, 1.509, -0.198, 0.011])(x)
     elseif x >  1 / 2.2
-        k = 2.659 * (1.040x - 1.857)
+        k = 2.659 * Poly([-1.857, 1.040])(x)
```
Comment (mileslucas) [inline src/color_laws.jl:132]: I realize I initially added the Polynomials here, but now that I think about it we really shouldn't be using them. Instantiating a polynomial requires an allocation, whereas using `@evalpoly x -2.156 1.508 -0.198 0.011`, for example, has no allocations.

File: src/color_laws.jl [hunk 6]
```diff
@@ -127,13 +127,87 @@ function cal00_invum(x::Real, Rv::Real)
...
+        a = 1.894 - 0.373 * x - 0.0101 / ((x - 4.57)^2 + 0.0384)
+        b = -3.490 + 2.057 * x + 0.706 / ((x - 4.59)^2 + 0.169)
+    else
+        return 0.0x
+    end
+    if 5.9 <= x <= 11.0  # far-NUV
+        y = x - 5.9
+        a += Poly([0.0, 0.0, -0.110, -0.0100])(y)
+        b += Poly([0.0, 0.0, 0.531, 0.0544])(y)
+    end
```
Comment (mileslucas) [inline src/color_laws.jl:173]: again, see my comments above on using `@evalpoly`

File: src/color_laws.jl [hunk 7]
```diff
@@ -127,13 +127,87 @@ function cal00_invum(x::Real, Rv::Real)
...
+    if 3.3 <= x <= 8.0  # NUV
+        a = 1.808 - 0.215 * x - 0.134 / ((x - 4.558)^2 + 0.566)
+        b = -2.350 + 1.403 * x + 1.103 / ((x - 4.587)^2 + 0.263)
+    else
+        return 0.0x
+    end
+    if 5.9 <= x <= 8.0  # far-NUV
+        y = x - 5.9
+        a += Poly([0.0, 0.0, -0.0077, -0.0030])(y)
+        b += Poly([0.0, 0.0, 0.2060, 0.0550])(y)
```
Comment (mileslucas) [inline src/color_laws.jl:209]: same thing, again.

File: test/color_laws.jl [hunk 8]
```diff
@@ -158,6 +158,85 @@ end
...
+
+    @test @inferred(broadcast(vcg04, wave)) ≈ ref_values[3.1] rtol = 0.016
+    @test_deprecated vcg04(wave)
+
+    for rv in [2.0, 3.0, 3.1, 4.0, 5.0, 6.0]
+        output = @inferred broadcast(vcg04, wave, rv)
+        @test output ≈ ref_values[rv] rtol = 0.016
+
+        # Test deprecated array syntax
+        @test_deprecated vcg04(wave, rv)
```
Comment (mileslucas) [inline test/color_laws.jl:185]: Don't test this if you remove the deprecation warning.

File: src/color_laws.jl (hunk 6 above)
Comment (siddharthlal25) [reply src/color_laws.jl]: updated them to `@evalpoly`!

File: src/color_laws.jl (hunk 7 above)
Comment (siddharthlal25) [reply src/color_laws.jl]: done!

File: test/color_laws.jl (hunk 8 above)
Comment (siddharthlal25) [reply test/color_laws.jl]: Removed the tests.

File: src/color_laws.jl (hunk 5 above)
Comment (siddharthlal25) [reply src/color_laws.jl]: Updated all to `@evalpoly` except in `ccm89_invum`.

File: src/DustExtinction.jl (hunk 4 above)
Comment (siddharthlal25) [reply src/DustExtinction.jl]: Done!

**Comment by giordano** [comment]
Cirrus is constantly failing across all projects I watch, I think we should just remove it...

**Comment by mileslucas** [comment]
Looks good as far as I can tell too. 

I guess let’s merge this and fix CI in another PR

**Comment by giordano** [comment]
Shall we merge this?

**Comment by siddharthlal25** [comment]
This PR is complete from my side. :smiley:

**Pull request merged**